eth_p2p/rlpx_protocols/shh_protocol.nim

```diff
@@ -416,31 +539,406 @@ proc add*(self: var Queue, msg: Message) =
+    for peer in node.peers(shh):
+      if not peer.state(shh).initialized:
+        continue
+
+      # NOTE: Perhaps alter the queue prune call to keep track of a HashSet
+      # of pruned messages (as these should be smaller), and diff this with
+      # the received sets.
+      peer.state(shh).received = intersection(peer.state(shh).received,
+                                              node.protocolState(shh).queue.itemHashes)
+
+proc run(node: EthereumNode, network: WhisperState) {.async.} =
+  while true:
+    # prune message queue every second
+    # TTL unit is in seconds, so this should be sufficient?
+    network.queue.prune()
+    # pruning the received sets is not necessary for correct workings
+    # but simply from keeping the sets growing indefinitely
+    node.pruneReceived()
+    await sleepAsync(1000)
+
+# Public EthereumNode calls ----------------------------------------------------
+
+proc sendP2PMessage*(node: EthereumNode, peerId: NodeId, env: Envelope): bool =
+  for peer in node.peers(shh):
+    if peer.remote.id == peerId:
+      asyncCheck peer.p2pMessage(env)
+      return true
+
+proc sendMessage*(node: EthereumNode, env: var Envelope): bool =
+  if not env.valid(): # actually just ttl !=0 is sufficient
+    return false
+
+  # We have to do the same checks here as in the messages proc not to leak
+  # any information that the message originates from this node.
+  let msg = initMessage(env)
+  if not msg.allowed(node.protocolState(shh).config):
+    return false
+
+  debug "Adding message to queue"
+  if node.protocolState(shh).queue.add(msg):
+    # Also notify our own filters of the message we are sending,
+    # e.g. msg from local Dapp to Dapp
+    node.protocolState(shh).filters.notify(msg)
+
+  return true
+
+proc postMessage*(node: EthereumNode, pubKey = none[PublicKey](),
+                  symKey = none[SymKey](), src = none[PrivateKey](),
+                  ttl: uint32, topic: Topic, payload: Bytes,
+                  padding = none[Bytes](), powTime = 1'f,
+                  targetPeer = none[NodeId]()): bool =
+  # NOTE: Allow a post without a key? Encryption is mandatory in v6?
+  let payload = encode(Payload(payload: payload, src: src, dst: pubKey,
+                               symKey: symKey, padding: padding))
+  if payload.isSome():
+    var env = Envelope(expiry:epochTime().uint32 + ttl + powTime.uint32,
+                       ttl: ttl, topic: topic, data: payload.get(), nonce: 0)
+
+    # Allow lightnode to post only direct p2p messages
+    if targetPeer.isSome():
+      return node.sendP2PMessage(targetPeer.get(), env)
+    elif not node.protocolState(shh).config.isLightNode:
+      # XXX: make this non blocking or not?
+      # In its current blocking state, it could be noticed by a peer that no
+      # messages are send for a while, and thus that mining PoW is done, and that
+      # next messages contains a message originated from this peer
+      env.nonce = env.minePow(powTime)
```

Comment:
It would be hard to execute this in a background thread at the moment. We'll need a way to send custom "tasks" to the async message loop (e.g. similar to C++'s `asio::post`). In `asyncdispatch2`, this is probably going to be represented by an `AsyncChannel` object.

Reply:
Yes, and I think that geth doesn't do this either (from seeing its behaviour when asking high PoW, I didn't check in the code yet).